forc-test/src/lib.rs

```diff
@@ -134,23 +143,21 @@ struct TestSetup {
 impl BuiltTests {
     /// Constructs a `PackageTests` from `Built`.
     ///
-    /// If the `built` is a workspace, `opts` is patched for the members of the workspace that
-    /// is a `Contract` so that only that contract is re-built.
-    ///
-    /// If the `built` is a package, `PackageTests::from_built_pkg` is used.
-    pub(crate) fn from_built(built: Built, opts: Opts) -> anyhow::Result<BuiltTests> {
+    /// Contracts are already compiled once without tests included to do `CONTRACT_ID` injection. `built_contracts` map holds already compiled contracts so that they can be matched with their "tests included" version.
+    pub(crate) fn from_built(
+        built: Built,
+        built_contracts: HashMap<String, BuiltPackage>,
```

Comment:
After realizing cargo does not allow package name duplication based on package name for the build plan, I opened-up #3840 and since duplicate names would not be possible in our build plan too, I though we could use String as key but if you feel like `Pinned` would be more robust here, happy to go that way too 🙌

Reply:
I ended up adding a field (`BuiltPackageDescriptor`) to `BuiltPackage` so that we can retrieve some information about the source of the package from `BuiltPackage`. This enabled me to use `Pinned` as a key as I have a way to get `Pinned` of a pkg built with tests and the `Pinned` of a pkg built with tests and built without tests is the same (since they are the same package).